### ISSUE TITLE
Job time not automatically tracked in user defined jobs

### ISSUE THREAD

**Issue by pmrv** [pyiron/pyiron_base#979]
@JNmpi discovered that the `timestart`/ `timestope` is not automatically tracked when user jobs defined do not explicitly call `self.db.update_item(self._runtime(), self.id)` or some such.

**Comment by jan-janssen** [comment]
As a first step I started to create a simpler `run_time_to_db()` function. I guess the easiest way is to define a decorator which wraps the run() function. https://github.com/pyiron/pyiron_base/pull/984

**Comment by pmrv** [comment]
I was actually thinking about setting `timestart` and `timestop` when entering and leaving status `running`.  That might take some refactoring, so a decorator in the mean sounds fine to me.

**Comment by jan-janssen** [comment]
> I was actually thinking about setting `timestart` and `timestop` when entering and leaving status `running`. That might take some refactoring, so a decorator in the mean sounds fine to me.

The issue is that the interactive jobs as well as the Genericmasters both call run_static multiple times.

**Comment by jan-janssen** [comment]
A decorator to track the run time is implemented in https://github.com/pyiron/pyiron_base/pull/992 still we might want to extend this for the Python Template Job, if and call `to_hdf()` also within the decorator.

**Comment by samwaseda** [comment]
> The issue is that the interactive jobs as well as the Genericmasters both call run_static multiple times.

Don't interactive jobs use `run_interactive`?

**Comment by samwaseda** [comment]
But actually I don't see the problem with calling `run_static` multiple times. Can't we simply update the run time whenever it's called? Basically, the starting time can be set only once (and therefore if something like [run decorator](https://github.com/pyiron/pyiron_base/pull/992) is called multiple times, the starting time wouldn't be updated), while the ending time is updated whenever the job sends the signal. In this regard, when `timestart` is set, it should also reinitialise `timestop` and `totalcputime`, i.e. `job.project.db.item_update({"timestart": datetime.now(), "timestop": None, "totalcputime": None}, job.job_id)`

**Comment by pmrv** [comment]
> > I was actually thinking about setting `timestart` and `timestop` when entering and leaving status `running`. That might take some refactoring, so a decorator in the mean sounds fine to me.
> 
> The issue is that the interactive jobs as well as the Genericmasters both call run_static multiple times.

That would be an issue for decorating `run_static`, but doing it on status change would still work, assuming these jobs do not set their state to `finished` multiple times.

**Comment by pmrv** [comment]
> A decorator to track the run time is implemented in #992 still we might want to extend this for the Python Template Job, if and call `to_hdf()` also within the decorator.

The decorator should only do one thing imo.

**Comment by samwaseda** [comment]
> That would be an issue for decorating `run_static`, but doing it on status change would still work, assuming these jobs do not set their state to `finished` multiple times.

Aha that means we set the starting time when `job.status.initialized = True` is set and ending time when `job.status.finished = True` (and maybe also `job.status.aborted = True`)? That sounds good to me.

**Comment by pmrv** [comment]
> > That would be an issue for decorating `run_static`, but doing it on status change would still work, assuming these jobs do not set their state to `finished` multiple times.
> 
> Aha that means we set the starting time when `job.status.initialized = True` is set and ending time when `job.status.finished = True` (and maybe also `job.status.aborted = True`)? That sounds good to me.

I was thinking to do it when the status is first set to `running` and then again when it is overwritten by any other status.

**Comment by samwaseda** [comment]
https://github.com/pyiron/pyiron_base/blob/d1a837261b4be36340ce4fa4778b921eabc4c77d/pyiron_base/jobs/job/extension/jobstatus.py#L252-L261

Ok that means in the lines above we add:

```python
if self.job_id is not None and self.database is not None:
    if name in job_status_finished_lst:
        totalcputime = datetime.now() - self.database.get_item_by_id(self.job_id)["timestart"]
        self.database.item_update({"timestop": datetime.now(), "totalcputime": totalcputime,}, self.job_id)
    elif name == 'running':
        self.database.item_update({"timestart": datetime.now()}, self.job_id)
```

Something like this?

**Comment by pmrv** [comment]
Yeah, maybe add a safeguard in case `timestart` is not set due to some inconsistencies.

**Comment by samwaseda** [comment]
Final form:

```python
if self.job_id is not None and self.database is not None:
    if name in job_status_finished_lst:
        timestart = self.database.get_item_by_id(self.job_id)["timestart"]
        time_dict = {"timestop": datetime.now()}
        if timestart is not None:
            totalcputime = datetime.now() - timestart
            time_dict["totalcputime"] = totalcputime
        else:
            warnings.warn("Starting time was not registered - total CPU time not registered")
        self.database.item_update(time_dict, self.job_id)
    elif name == 'running':
        self.database.item_update({"timestart": datetime.now()}, self.job_id)
```

If there's someone who can open a PR with this change, I'd be grateful, because right now I'm already running so many jobs on the cluster and I cannot test it with job submission.

**Comment by samwaseda** [comment]
And I remotely remember that `warnings` was not the right way to give warnings in pyiron anymore, but I cannot remember how to do it correctly. Sounds like sort of a thing that should be documented somewhere.

**Comment by pmrv** [comment]
I'll see if I get to it in the afternoon.

**Comment by jan-janssen** [comment]
> Something like this?

I am not sure if we want to add the run time calculation to the `jobstatus` class, it should only focus on setting the status and syncing with the database. That is why my suggestion was to add a decorator around the `run_static()` function. Then its also clear how people can modify their function to add this functionality.

**Comment by pmrv** [comment]
Is there a case where you'd not want to have the run time tracked though?

**Comment by samwaseda** [comment]
> I am not sure if we want to add the run time calculation to the `jobstatus` class, it should only focus on setting the status and syncing with the database. That is why my suggestion was to add a decorator around the `run_static()` function. Then its also clear how people can modify their function to add this functionality.

I see your point and I'm also slightly uncomfortable with the setting the CPU time inside `JobStatus`. However, at least the current reality is that the CPU time is not set consistently. And even if the decorator is ready, it will have to be applied in various places. With @pmrv's idea, I have the feeling that we can totally unify the CPU time setting in one place in a very unambiguous way. So I might still prefer his suggestion.

**Comment by jan-janssen** [comment]
> Is there a case where you'd not want to have the run time tracked though?

It is more that I want one class for one purpose. Handling the runtime to me is something different than handling the job status.

**Comment by jan-janssen** [comment]
> I see your point and I'm also slightly uncomfortable with the setting the CPU time inside `JobStatus`. However, at least the current reality is that the CPU time is not set consistently. And even if the decorator is ready, it will have to be applied in various places. With @pmrv's idea, I have the feeling that we can totally unify the CPU time setting in one place in a very unambiguous way. So I might still prefer his suggestion.

As I said, the `GenericMasters` use `run_static()` as well. Here the `run_static()` is called multiple times, as the job is basically iterating between status `running` and `suspended` . To minimize the changes and first focus on simplifying the implementation of new simulation codes. Once we fully understand the challenges and managed the transition to the Datacontainer then it is hopefully easier to clean up the run function.
